### PR TITLE
fix(xds-server) log network interface

### LINKED ISSUE #408: log network interface and port for each server that `kuma-cp` starts
### Summary

* at the moment, the log of `kuma-cp` looks like this

  ```
  2019-11-06T08:17:38.952+0100	INFO	Skipping reading config from file
  2019-11-06T08:17:39.066+0100	INFO	bootstrap.auto-configure	auto-generated TLS certificate for SDS server	{"crtFile": "/var/folders/rt/6pw2xrn53q91d4qmr595y24r0000gp/T/672876359.crt", "keyFile": "/var/folders/rt/6pw2xrn53q91d4qmr595y24r0000gp/T/107755514.key"}
  2019-11-06T08:17:39.066+0100	INFO	bootstrap	Creating default mesh from the settings	 
  {"mesh": {"mtls":{"ca":{"Type":null}}}}
  2019-11-06T08:17:39.355+0100	INFO	kuma-cp.run	starting Control Plane
  2019-11-06T08:17:39.355+0100	INFO	api-server	starting	{"port": ":5681"}
  2019-11-06T08:17:39.355+0100	INFO	xds-server.diagnostics	starting	{"port": 5680}
  2019-11-06T08:17:39.355+0100	INFO	bootstrap-server	starting	{"port": 5682}
  2019-11-06T08:17:39.355+0100	INFO	dataplane-token-server	starting server	{"port": 5679}
  2019-11-06T08:17:39.355+0100	INFO	xds-server.grpc	starting	{"port": 5678}
  2019-11-06T08:17:39.356+0100	INFO	sds-server.grpc	starting	{"port": 5677, "tls": true}
  ```
* it already includes `port` for each network server, but not `interface`

### Expected behaviour

* add `interface` param to the log, e.g.

  ```
  2019-11-06T08:17:39.355+0100	INFO	xds-server.grpc	starting	{"interface": "0.0.0.0", "port": 5678}
  ```

### PR DESCRIPTION
**Pull request by tharun208**
* Added logging of interface

Fix #408

### Issues resolved

Fix #408

### CODE REVIEW

**Comment by jakubdyszkiewicz** [comment]
Hey, @tharun208 thanks for the contribution!

Can you rebase to origin/master so the CI will pass?

Also, can you also add an interface to every server that kuma-cp exposes? You can see the servers in the logs when kuma-cp starts.

**Comment by tharun208** [comment]
> Hey, @tharun208 thanks for the contribution!
> 
> Can you rebase to origin/master so the CI will pass?
> 
> Also, can you also add an interface to every server that kuma-cp exposes? You can see the servers in the logs when kuma-cp starts.

Hi @jakubdyszkiewicz , Need clarification on the interface, `net` Address is a complete address with the hostname and port number. Should we need to split it?

**Comment by tharun208** [comment]
> Hey, @tharun208 thanks for the contribution!
> 
> Can you rebase to origin/master so the CI will pass?
> It is up to date, clang-format on ubuntu image is failing

**Comment by jakubdyszkiewicz** [comment]
> Hi @jakubdyszkiewicz , Need clarification on the interface, net Address is a complete address with the hostname and port number. Should we need to split it?

In this case, I'd go with `{"address": "127.0.0.1:5681"}` instead of `{"interface": "127.0.0.1", "port": 5681}`

**Comment by tharun208** [comment]
> > Hi @jakubdyszkiewicz , Need clarification on the interface, net Address is a complete address with the hostname and port number. Should we need to split it?
> 
> In this case, I'd go with `{"address": "127.0.0.1:5681"}` instead of `{"interface": "127.0.0.1", "port": 5681}`

Cool, Will Update and make a PR :)

**Comment by tharun208** [comment]
> > Hey, @tharun208 thanks for the contribution!
> > Can you rebase to origin/master so the CI will pass?
> > Also, can you also add an interface to every server that kuma-cp exposes? You can see the servers in the logs when kuma-cp starts.
> 
> Hi @jakubdyszkiewicz , Need clarification on the interface, `net` Address is a complete address with the hostname and port number. Should we need to split it?

Is the CI issue resolved?

**Comment by jakubdyszkiewicz** [comment]
@tharun208 the issue is resolved but the tests are failing because you've changed the behaviour.

`fmt.Sprintf(":%d", g.Config.Port)` means that the server is exposed on this port on every interface (other notation is `0.0.0.0:port`) while `fmt.Sprintf("127.0.0.1:%d", g.Config.Port)` means that server is exposed only on `localhost:port` so you cannot access the control plane from any other machine.

We should log `{"address": "0.0.0.0:5678"}` (or we can go back to `{"interface": "0.0.0.0", "port": 5678}` if you wish). All servers are exposed on `0.0.0.0` except of the admin-server.

**Comment by tharun208** [comment]
I changed the tests and pushed but the ci is failing on minikube build

**Comment by jakubdyszkiewicz** [comment]
@tharun208 please take a look at my last comment. Let me know if you need clarification on that.

File: pkg/admin-server/server.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ func (a *AdminServer) Start(stop <-chan struct{}) error {
 
 func (a *AdminServer) startHttpServer() (*http.Server, chan error) {
 	server := &http.Server{
-		Addr:    fmt.Sprintf("127.0.0.1:%d", a.cfg.Local.Port),
+		Addr:    fmt.Sprintf("%s:%d", a.cfg.Public.Interface, a.cfg.Local.Port),
```
Comment (jakubdyszkiewicz) [inline pkg/admin-server/server.go]: HTTP admin server is the only one that should run on `127.0.0.1`, so hardcode here was fine

File: pkg/mads/server/grpc.go [hunk 2]
```diff
@@ -34,7 +34,7 @@ func (s *grpcServer) Start(stop <-chan struct{}) error {
 	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
 	grpcServer := grpc.NewServer(grpcOptions...)
 
-	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", s.config.GrpcPort))
+	lis, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", s.config.GrpcPort))
```
Comment (jakubdyszkiewicz) [inline pkg/mads/server/grpc.go]: Please revert this back so the server can be accessed outside of the control plane machine.

File: pkg/sds/server/grpc.go [hunk 3]
```diff
@@ -44,7 +44,7 @@ func (s *grpcServer) Start(stop <-chan struct{}) error {
 	}
 	grpcServer := grpc.NewServer(grpcOptions...)
 
-	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", s.config.GrpcPort))
+	lis, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", s.config.GrpcPort))
```
Comment (jakubdyszkiewicz) [inline pkg/sds/server/grpc.go]: Please revert this back so the server can be accessed outside of the control plane machine.

File: pkg/xds/bootstrap/server.go [hunk 4]
```diff
@@ -28,7 +28,7 @@ func (b *BootstrapServer) Start(stop <-chan struct{}) error {
 	mux.HandleFunc("/bootstrap", b.handleBootstrapRequest)
 
 	bootstrapServer := &http.Server{
-		Addr:    fmt.Sprintf(":%d", b.Port),
+		Addr:    fmt.Sprintf("127.0.0.1:%d", b.Port),
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/server.go]: Please revert this back so the server can be accessed outside of the control plane machine.

File: pkg/xds/server/diagnostics.go [hunk 5]
```diff
@@ -31,7 +31,7 @@ func (s *diagnosticsServer) Start(stop <-chan struct{}) error {
 		resp.WriteHeader(http.StatusOK)
 	})
 
-	httpServer := &http.Server{Addr: fmt.Sprintf(":%d", s.port), Handler: mux}
+	httpServer := &http.Server{Addr: fmt.Sprintf("127.0.0.1:%d", s.port), Handler: mux}
```
Comment (jakubdyszkiewicz) [inline pkg/xds/server/diagnostics.go]: Please revert this back so the server can be accessed outside of the control plane machine.

File: pkg/xds/server/grpc.go [hunk 6]
```diff
@@ -33,7 +33,7 @@ func (s *grpcServer) Start(stop <-chan struct{}) error {
 	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
 	grpcServer := grpc.NewServer(grpcOptions...)
 
-	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", s.port))
+	lis, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", s.port))
```
Comment (jakubdyszkiewicz) [inline pkg/xds/server/grpc.go]: Please revert this back so the server can be accessed outside of the control plane machine.

**Comment by devadvocado** [comment]
Hey @tharun208, huge thanks to getting this through. Just have to ask you one last thing: rebase to origin/master so the CI will pass. I see your branch doesn't have [Jakub's fix to the clang issue](https://github.com/Kong/kuma/commit/b352accc9b5981062eae4cfa3e79c90a378318ce).

**Comment by tharun208** [comment]
> Jakub's fix to the clang issue

It is up to date saying when I did git rebase master.

**Comment by devadvocado** [comment]
@tharun208 it should be but unfortunately I don't see any of the recent commits in your branch, which the rebase should've fixed. would you kindly rebase again to master? Thanks!

**Comment by jakubdyszkiewicz** [comment]
@tharun208 make sure that you rebase to updated `origin/master`. Your local `master` may be outdated.

For current changes. Right now the logs look like this
```
2020-02-04T13:40:45.818+0100	INFO	admin-server	starting server	{"address": "127.0.0.1:5679"}
2020-02-04T13:40:45.818+0100	INFO	gui-server	starting	{"address": ":5683"}
2020-02-04T13:40:45.818+0100	INFO	api-server	starting	{"address": ":5681"}
2020-02-04T13:40:45.818+0100	INFO	bootstrap-server	starting	{"address": ":5682"}
2020-02-04T13:40:45.818+0100	INFO	xds-server.diagnostics	starting	{"address": ":5680"}
2020-02-04T13:40:45.819+0100	INFO	mads-server.grpc	starting	{"address": "[::]:5676"}
2020-02-04T13:40:45.819+0100	INFO	xds-server.grpc	starting	{"address": "[::]:5678"}
2020-02-04T13:40:45.819+0100	INFO	sds-server.grpc	starting	{"address": "[::]:5677", "tls": true}
```

This :5681 notation might be a little bit confusing for non golang devs that may go through Kuma logs therefore we would like to see it like this

```
2020-02-04T13:40:45.818+0100	INFO	admin-server	starting server	{"address": "127.0.0.1:5679"}
2020-02-04T13:40:45.818+0100	INFO	gui-server	starting	{"address": "0.0.0.0:5683"}
2020-02-04T13:40:45.818+0100	INFO	api-server	starting	{"address": "0.0.0.0:5681"}
2020-02-04T13:40:45.818+0100	INFO	bootstrap-server	starting	{"address": "0.0.0.0:5682"}
2020-02-04T13:40:45.818+0100	INFO	xds-server.diagnostics	starting	{"address": "0.0.0.0:5680"}
2020-02-04T13:40:45.819+0100	INFO	mads-server.grpc	starting	{"address": "0.0.0.0:5676"}
2020-02-04T13:40:45.819+0100	INFO	xds-server.grpc	starting	{"address": "0.0.0.0:5678"}
2020-02-04T13:40:45.819+0100	INFO	sds-server.grpc	starting	{"address": "0.0.0.0:5677", "tls": true}
```

or

```
2020-02-04T13:40:45.818+0100	INFO	admin-server	starting server	{"interface": "127.0.0.1", "port": 5679}
2020-02-04T13:40:45.818+0100	INFO	gui-server	starting	{"interface": "0.0.0.0", "port": 5683}
2020-02-04T13:40:45.818+0100	INFO	api-server	starting	{"interface": "0.0.0.0", "port": 5681}
2020-02-04T13:40:45.818+0100	INFO	bootstrap-server	starting	{"interface": "0.0.0.0", "port": 5682}
2020-02-04T13:40:45.818+0100	INFO	xds-server.diagnostics	starting	{"interface": "0.0.0.0", "port": 5680}
2020-02-04T13:40:45.819+0100	INFO	mads-server.grpc	starting	{"interface": "0.0.0.0", "port": 5676}
2020-02-04T13:40:45.819+0100	INFO	xds-server.grpc	starting	{"interface": "0.0.0.0", "port": 5678}
2020-02-04T13:40:45.819+0100	INFO	sds-server.grpc	starting	{"interface": "0.0.0.0", "port": 5677, "tls": true}
```

Thanks

**Comment by tharun208** [comment]
Hi, @jakubdyszkiewicz  @devadvocado Rebased and good to go now :)

**Comment by jakubdyszkiewicz** [comment]
Hey, thanks but I see `+2,514 −1,679` changes as well as 21 commits.
I'd suggest to do `git reset origin/master` (not hard, soft), and then commit changes.

**Pull request closed without merging**